plugins/ansilexer.py

```diff
@@ -28,83 +29,288 @@
-        yield (match.start, string_to_tokentype(token), match.group(4))
+    def _to_hex(self, r, g, b):
+        return '{:02x}{:02x}{:02x}'.format(r, g, b)
 
     tokens = {
         'root': [
             ('[^\x1b]+', Text),
-            ('\x1b\\[38;2;(\\d+);(\\d+);(\\d+)m\x1b\\[48;2;\\d+;\\d+;\\d+m([^\x1b]+)\x1b\\[0m', callback_fg_bg_color),
-            ('\x1b\\[38;2;(\\d+);(\\d+);(\\d+)m([^\x1b]+)\x1b\\[0m', callback_fg_color),
-            # Brightness and color separately (used by ASan reports)
-            ('\x1b\\[([01])m\x1b\\[(\\d+)m([^\x1b]*)', callback),
-            ('\x1b\\[(\\d+)(?:;(\\d+))?m([^\x1b]*)', callback)
+            (_sgrs_and_text, _callback),
         ]
     }
 
-_ansi_fg_color_re = re.compile('class="g g-AnsiForegroundColor([0-9a-f]{6})">')
-_ansi_fg_bg_color_re = re.compile('class="g g-AnsiForegroundBackgroundColor([0-9a-f]{6})">')
-
 class HtmlAnsiFormatter(HtmlFormatter):
+    _ansi_color_re = re.compile(
+        r'(?P<Prefix>class="g )(?P<AnsiClass>g-Ansi\S+)(?P<Suffix>">)'
+    )
+    _rgb_code_re = re.compile(r'[0-9a-f]{6}$')
+
     def wrap(self, source, outfile):
         return self._wrap_code(source)
 
     def _wrap_code(self, source):
         for i, t in source:
             if i == 1: # it's a line of formatted code
-                # Add ZWNJ for before each character because otherwise it's
-                # somehow impossible to wrap even with word-break: break-all.
-                # Not sure why (and not sure if this is the best solution), but
-                # had to ship a thing so there it is. Adding <wbr/> did not
-                # help.
-                t = _ansi_fg_bg_color_re.sub('style="color: #\\1; background-color: #\\1">&zwnj;', t)
-                t = _ansi_fg_color_re.sub('style="color: #\\1">&zwnj;', t)
-                #t += 'H'
+                t = self._ansi_color_re.sub(self._replace_ansi_class, t)
             yield i, t
+
+    def _replace_ansi_class(self, match):
+        html_classes = ['g']
+        html_styles = []
+        foreground_color = None
+        background_color = None
+        ansi_class = match.group('AnsiClass')
+
+        i = ansi_class.find('BackgroundColor')
+        if i >= 0:
+            background_color = ansi_class[i + 15:]
+            ansi_class = ansi_class[:i]
+
+            if self._rgb_code_re.match(background_color):
+                html_styles.append('background-color: #' + background_color)
+            else:
+                html_classes.append('g-AnsiBackground' + background_color)
+
+        i = ansi_class.find('ForegroundColor')
+        if i >= 0:
+            foreground_color = ansi_class[i + 15:]
+            ansi_class = ansi_class[:i]
+            html_styles.insert(0, 'color: #' + foreground_color)
+
+        if ansi_class != 'g-Ansi':
+            html_classes.insert(1, ansi_class)
+
+        result = ''
+        if len(html_classes) > 1: # we don't want to emit just g
```

Comment:
Previously, you'd dropped the `g` when emitting an RGB-styled foreground color via the style attribute, so I assumed this was intentional and drop it whenever there's no class that contributes to the colour.